### PR TITLE
feat(nm): Wifi status retrieval support for New Kura Networking

### PR DESCRIPTION
**Pull request by mattdibi**
**Related Issue:** https://github.com/eclipse/kura/pull/4363

**Description of the solution adopted:** This PR adds the required code to retrieve WiFi status informations from the NetworkManager DBus API.

#### Classes for representing NetworkManager D-Bus API Types
- Added `NM80211Mode` and related tests for representing: https://networkmanager.dev/docs/api/latest/nm-dbus-types.html#NM80211Mode
- Added `NM80211ApSecurityFlags` and related tests for representing: https://networkmanager.dev/docs/api/latest/nm-dbus-types.html#NM80211ApSecurityFlags
- Added `NNMDeviceWifiCapabilities` and related tests for representing: https://networkmanager.dev/docs/api/latest/nm-dbus-types.html#NMDeviceWifiCapabilities

#### Data structures
The following classes were added just to bundle together closely related informations retrieved from the DBus API and make Sonar happy about the number of parameter passed to methods.
- `WirelessProperties`
- `SupportedChannelsProperties`
- `AccessPointsProperties`

#### NMDbusConnector

NMDbusConnector is responsible for retrieving the information from NetworkManager and `iw`. To retrieve the regulatory domain and the supported channels for a network interface we decided to use the `IwCapabilityTool` from `org.eclipse.kura.linux.net.util`, thus the dependency on `iw`.

As we know this is not ideal, but we'll leave it as-is for now and hopefully use another mechanism for retrieving this information in the future.

#### NMStatusConverter

Is responsible for converting the data retrieved from NetworkManager into the newly introduced Network Status information data structures: https://github.com/eclipse/kura/pull/4378

**Note**: Unit tests are missing and thus the Sonar checks are failing (coverage < 50%). Will be added in a future PR.

### CODE REVIEW

**Comment by pierantoniomerlino** [comment]
The `IwCapabilityTool` is in the `org.eclipse.kura.linux.net.util` package. So, the wifi status has a dependency from that bundle.

File: kura/org.eclipse.kura.nm/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -27,6 +28,7 @@ Import-Package: org.apache.commons.io;version="2.4.0",
  org.eclipse.kura.net.status.loopback;version="[1.0,2.0)",
  org.eclipse.kura.net.status.modem;version="[1.0,2.0)",
  org.eclipse.kura.net.status.wifi;version="[1.0,2.0)",
+ org.eclipse.kura.net.wifi;version="2.3.0",
  org.eclipse.kura.usb;version="[1.0,2.0)",
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/META-INF/MANIFEST.MF:32]: Why the import of 'org.eclipse.kura.net.wifi` is needed? The `IwCapabilityTool` is in `org.eclipse.kura.linux.net.util` that is already in the dependency list. BTW, if the package is needed, a range is required.

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/status/NMStatusConverter.java [hunk 2]
```diff
@@ -111,6 +151,242 @@ private static void setIP4Status(NetworkInterfaceStatusBuilder<?> builder,
...
+        NM80211Mode mode = NM80211Mode.fromUInt32(wirelessDeviceProperties.Get(NM_DEVICE_WIRELESS_BUS_NAME, "Mode"));
+        builder.withMode(wifiModeConvert(mode));
+
+        List<NMDeviceWifiCapabilities> capabilities = NMDeviceWifiCapabilities
+                .fromUInt32(wirelessDeviceProperties.Get(NM_DEVICE_WIRELESS_BUS_NAME, "WirelessCapabilities"));
+        builder.withCapabilities(wifiCapabilitiesConvert(capabilities));
+        builder.withSupportedRadioModes(wifiRadioModesConvert(capabilities));
+
+        // builder.withSupportedBitrates
+
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/status/NMStatusConverter.java]: Yeah, probably the `supportedBitrates` is useless and should be replaced with `MaxBitrate`. However, please remove the comment.

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/status/NMStatusConverter.java [hunk 3]
```diff
@@ -111,6 +151,242 @@ private static void setIP4Status(NetworkInterfaceStatusBuilder<?> builder,
...
+        Pair<List<Integer>, List<Long>> kuraSupportedChannels = wifiChannelsConvert(supportedChannels);
+        builder.withSupportedChannels(kuraSupportedChannels.getLeft());
+        builder.withSupportedFrequencies(kuraSupportedChannels.getRight());
+
+        builder.withCountryCode(countryCode);
+
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/status/NMStatusConverter.java:170]: Indentation?

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/status/NMStatusConverter.java [hunk 4]
```diff
@@ -111,6 +151,242 @@ private static void setIP4Status(NetworkInterfaceStatusBuilder<?> builder,
...
+            builder.withActiveWifiAccessPoint(Optional.of(ap));
+        } else {
+            if (activeAccessPoint.isPresent()) {
+                WifiAccessPoint ap = wifiAccessPointConvert(activeAccessPoint.get());
+                builder.withActiveWifiAccessPoint(Optional.of(ap));
+            }
+        }
+
+        builder.withAvailableWifiAccessPoints(wifiAccessPointConvert(accessPoints));
+    }
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/status/NMStatusConverter.java:187]: Does it make sense to add the available AP even if the interface is configured in MASTER mode?

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/status/NMStatusConverter.java (hunk 2 above)
Comment (mattdibi) [reply kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/status/NMStatusConverter.java]: Fixed in: 0464c511bbec589357d53f26c98a60d43ec702bf

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/status/NMStatusConverter.java (hunk 3 above)
Comment (mattdibi) [reply kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/status/NMStatusConverter.java:170]: I... don't see any issue here with the indentation... what's wrong with that?

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/status/NMStatusConverter.java (hunk 4 above)
Comment (mattdibi) [reply kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/status/NMStatusConverter.java:187]: IMO yes: if I want to change the interface from AP to infra mode I already have the available access point to connect to.

File: kura/org.eclipse.kura.nm/META-INF/MANIFEST.MF (hunk 1 above)
Comment (mattdibi) [reply kura/org.eclipse.kura.nm/META-INF/MANIFEST.MF:32]: Uh sorry, I got it wrong in the PR description.

I need `org.eclipse.kura.net.wifi` for the `WifiChannel` data structure which is used here: https://github.com/eclipse/kura/blob/b26338832c9857de3a77b60f4c5aa83e7341e5a6/kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java#L181

It is what the `IwCapabilityTool` returns for the supported channels. Will update with a range.

Comment (mattdibi) [reply kura/org.eclipse.kura.nm/META-INF/MANIFEST.MF:32]: Fixed in: 80e3a901425f56f431a5b09536944560340f3310

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/status/NMStatusConverter.java (hunk 3 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/status/NMStatusConverter.java:170]: No, it's ok. My fault.

**Review by pierantoniomerlino — APPROVED** [review]

**Comment by pierantoniomerlino** [comment]
The build succeeded, but the Sonar check fails for a too low code coverage. This will be increased by #4408. 
Moreover, the Sonar report is not uploaded for an issue in the CI.

**Pull request merged**